### PR TITLE
Support multi-name combinatoric queries

### PR DESCRIPTION
**Pull request by tomthorogood**
This achieves parity+ for name searched compared to the whitepages-ui.

-  Expands model validation for phone number and name.
- Adds services/query_generator.py
- Updates services/search.py to use the query generator
- Updates test_app and test_search
- Adds test_query_generator

### CODE REVIEW

File: husky_directory/app.py [hunk 1]
```diff
@@ -100,7 +100,11 @@ def dispatch_request(self):
         request_input = self.injector.get(SearchDirectoryInput)
         self.logger.info(f"searching for {request_input}")
         request_output = self.directory.search_directory(request_input)
-        return jsonify(request_output.dict(by_alias=True))
+        result = {
```
Comment (goulter) [inline husky_directory/app.py:103]: What does the result object contain after line 105 runs?

Comment (tomthorogood) [reply husky_directory/app.py:103]:
```
{
   "query description x": {"people": [ {person...}, {person...}, {person...}]},
   "query description y": {"people": [ {person...}, {person...}, {person...}]},
   ...
}
```

Comment (tomthorogood) [reply husky_directory/app.py:103]: Screen Shot 2021-01-08 at 9 33 29 AM

Comment (tomthorogood) [reply husky_directory/app.py:103]: (for now; I'll probably change this into a more explicit model once I have the other searches working and feel like the underlying models are solidified)

Comment (goulter) [reply husky_directory/app.py:103]: Ah, so 105 returns a `descr` and `result` for each "item" in `request_output`?

Comment (goulter) [reply husky_directory/app.py:103]: I'm not second guessing your code, just upping my understanding 👍

Comment (tomthorogood) [reply husky_directory/app.py:103]: yes, this is known as "dictionary comprehension." 

the end result is the same as if i had written:

```
results = {}
for description, result in request_output.items():
    results[description] = result.dict(by_alias=True)
```

Dictionary comprehensions (and list/set comprehensions) are efficient built-in python constructs.

File: husky_directory/services/query_generator.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+
+Query = ListPersonsInput  # Aliasing this for brevity in the below definitions.
+ArgFmt = WildcardFormat  # An uglier, but shorter name
+
+
+# TODO: Once auth is enabled and set up:
+#       also include netid queries for authenticated users
+#       https://github.com/UWIT-IAM/whitepages-ui/blob/master/wpquery.c#L426-L434
+#  (Also, create a jira for this, and remove this comment.)
```
Comment (tomthorogood) [inline husky_directory/services/query_generator.py:96]: Replacing this with EDS-514

File: husky_directory/services/query_generator.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            return
+
+        for template in self._generate_sliced_name_queries(cardinality):
+            yield template.get_description(name_parts), template.get_query(name_parts)
+
+    def generate(
+        self, request_input: SearchDirectoryInput
+    ) -> Iterable[Tuple[str, ListPersonsInput]]:
+        # TODO: generate queries for other field types (phone, box number, email)
+        #       Those will all be much easier because we don't have to contend with combinatorics as much.
```
Comment (tomthorogood) [inline husky_directory/services/query_generator.py:269]: Replacing with EDS-515, EDS-516, EDS-517

**Review by goulter — APPROVED** [review]

**Pull request merged**